code/ai/aigoals.cpp

```diff
@@ -2349,14 +2433,32 @@ void ai_process_mission_orders( int objnum, ai_info *aip )
 	}	
 
 
+	// save the current goal (if any) first, in case it's wiped out by the next action
```

Comment:
After we check for `Clear_all_goals_first` below, we shouldn't refer to `current_goal` again, right?

So there's more data from `current_goal` we need to save at this point, right?

Although I'm a bit confused. Since we were clearing goals before, why do we need to backup data from `current_goal`? Why wasn't this necessary before?

Reply:
On 2nd thought, instead of saving individual fields from `current_goal`, here's what I suggest:
1. Keep an extra `ai_goal` instance in memory called, say, `current_goal_backup`.
2. Right before calling `ai_clear_ship_goals()` below, first copy the contents of `*current_goal` to `current_goal_backup`, and set `current_goal` to be `&current_goal_backup`.

That way, `current_goal` will always be a valid pointer.

Reply:
This wasn't necessary before because the implementations of the relevant goals handled the clearing themselves. (In the case of form-on-wing it had a bug analogous to use-after-free, but it worked anyhow).  Since the clearing is now generalized, saving the data is now necessary.

Your idea of backing up the current goal is a really good one.  I'll implement it with a slight tweak.